cardano-cli/src/Cardano/CLI/Shelley/Commands.hs

```diff
@@ -477,6 +487,24 @@ renderGenesisCmd cmd =
 -- Shelley CLI flag/option data types
 --
 
+data NetworkIdArg
+  = NetworkIdFromCLI NetworkId
+  | NetworkIdFromEnv
+  deriving (Show,Eq)
+
+-- Commands with this argument
+-- behave different depending on whether a network-id option is given or not.
+--
+data SignForNetworkId
+  = SignForNetworkId !NetworkId
+  | TakeNetworkIdFromKey
```

Comment:
Instead of creating a new sum type and new parsers why not just use `Maybe NetworkId` like before? `SignForNetworkId` doesn't really bring much value over `Maybe NetworkId`

Reply:
What you suggest is the situation as it is.
And this situation is very unsatisfactory because at the moment `Maybe NetworkId` is used with two different meanings and two different interpretations in the (existing) code.
Some of the commands resolve `Maybe NetworkId` using the network ID from the environment or the flag.
But there is also the one that takes the network ID from the key in the `Nothing` case.
The best way to handle that is to create a new data type in Haskell.
That is very cheap and helps to cleanup stuff a lot.
It is a common pattern to use
`data MyFeature = HaveFeature | DontHaveFeature`
instead of
`type MyFeature = Bool`

Reply:
Sure but I doubt it's worth further increasing the complexity + size of the codebase to differentiate between these two cases. We already have numerous data definitions in the cli and I don't want to introduce anymore unless the benefit is significant. I don't see that being the case here.

Reply:
It is much more readable, adds very little overhead and helps to avoid bugs.
When adding support for networkId from env it was necessary to differentiate between these two cases.
if done right (!) it can even help to reduce code complexity and code size.
Also can help to get better help messages in the CLI.

Reply:
[Here](https://github.com/input-output-hk/cardano-node/blob/022d552ce8f72f1a2cb19410bbb5243a4ec539a5/cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs#L1183) is an example.
`Maybe NetworkId` is defaulted to `Mainnet` right in the middle of some code.
With a proper datatype it would be easier to spot and avoid unwanted defaults.

Reply:
The way to fix that change the function type signature from: 
```
runTxCalculateMinFee
  :: TxBodyFile
  -> Maybe NetworkId
```
to 
```
runTxCalculateMinFee
  :: TxBodyFile
  -> NetworkId
```
Which _eliminates_ the need to add another data definition.